File: services/videoProcessor.ts
/**
 * Video Processor - Processamento de vídeo 100% no browser com FFmpeg.wasm
 * Sem necessidade de backend/servidor.
 */

import { FFmpeg } from "@ffmpeg/ffmpeg";
import { fetchFile, toBlobURL } from "@ffmpeg/util";
import type { CompositionSettings } from "../types";

let ffmpegInstance: FFmpeg | null = null;
let ffmpegLoading = false;
let ffmpegLoaded = false;

/**
 * Carrega e inicializa o FFmpeg.wasm
 */
export async function loadFFmpeg(
  onProgress?: (msg: string) => void
): Promise<FFmpeg> {
  if (ffmpegInstance && ffmpegLoaded) return ffmpegInstance;

  if (ffmpegLoading) {
    // Aguardar carregamento em andamento
    while (ffmpegLoading) {
      await new Promise((r) => setTimeout(r, 100));
    }
    if (ffmpegInstance && ffmpegLoaded) return ffmpegInstance;
  }

  ffmpegLoading = true;
  onProgress?.("Carregando motor de vídeo...");

  try {
    const ffmpeg = new FFmpeg();

    ffmpeg.on("log", ({ message }) => {
      // console.log("[FFmpeg]", message);
    });

    const baseURL = "https://unpkg.com/@ffmpeg/core@0.12.6/dist/esm";

    onProgress?.("Baixando FFmpeg WebAssembly...");

    await ffmpeg.load({
      coreURL: await toBlobURL(`${baseURL}/ffmpeg-core.js`, "text/javascript"),
      wasmURL: await toBlobURL(
        `${baseURL}/ffmpeg-core.wasm`,
        "application/wasm"
      ),
    });

    ffmpegInstance = ffmpeg;
    ffmpegLoaded = true;
    onProgress?.("Motor de vídeo pronto!");
    return ffmpeg;
  } finally {
    ffmpegLoading = false;
  }
}

/**
 * Obtém metadados de um arquivo de vídeo usando HTMLVideoElement
 */
export function getVideoMetadata(
  file: File
): Promise<{ duration: number; width: number; height: number }> {
  return new Promise((resolve, reject) => {
    const video = document.createElement("video");
    video.preload = "metadata";

    const url = URL.createObjectURL(file);

    video.onloadedmetadata = () => {
      resolve({
        duration: video.duration,
        width: video.videoWidth,
        height: video.videoHeight,
      });
      URL.revokeObjectURL(url);
    };

    video.onerror = () => {
      URL.revokeObjectURL(url);
      reject(new Error(`Não foi possível ler o vídeo: ${file.name}`));
    };

    video.src = url;
  });
}

/**
 * Gera uma thumbnail de um vídeo usando Canvas
 */
export function generateThumbnail(
  file: File,
  time: number = 0.5
): Promise<string> {
  return new Promise((resolve, reject) => {
    const video = document.createElement("video");
    video.preload = "auto";
    video.muted = true;
    video.playsInline = true;

    const url = URL.createObjectURL(file);

    video.onloadeddata = () => {
      video.currentTime = Math.min(time, video.duration * 0.3);
    };

    video.onseeked = () => {
      try {
        const canvas = document.createElement("canvas");
        const scale = 160 / Math.max(video.videoWidth, video.videoHeight);
        canvas.width = Math.round(video.videoWidth * scale);
        canvas.height = Math.round(video.videoHeight * scale);

        const ctx = canvas.getContext("2d");
        if (ctx) {
          ctx.drawImage(video, 0, 0, canvas.width, canvas.height);
          const dataUrl = canvas.toDataURL("image/jpeg", 0.7);
          resolve(dataUrl);
        } else {
          resolve("");
        }
      } catch {
        resolve("");
      } finally {
        URL.revokeObjectURL(url);
      }
    };

    video.onerror = () => {
      URL.revokeObjectURL(url);
      resolve(""); // Não falhar por causa de thumbnail
    };

    video.src = url;
  });
}

/**
 * Calcula as durações dos clipes para atingir a duração alvo
 */
function calculateClipDurations(
  originalDurations: number[],
  targetDuration: number,
  transitionDuration: number
): number[] {
  const n = originalDurations.length;
  if (n === 0) return [];

  const totalTransitionTime = (n - 1) * transitionDuration;
  const availableTime = targetDuration + totalTransitionTime;
  const totalOriginal = originalDurations.reduce((a, b) => a + b, 0);

  if (totalOriginal <= 0) {
    const perClip = availableTime / n;
    return originalDurations.map(() => perClip);
  }

  return originalDurations.map((dur) => {
    const ratio = dur / totalOriginal;
    let clipDur = ratio * availableTime;
    clipDur = Math.min(clipDur, dur);
    clipDur = Math.max(clipDur, 1.5);
    return clipDur;
  });
}

/**
 * Mapeamento de nomes de transição para o FFmpeg xfade
 */
const TRANSITION_MAP: Record<string, string> = {
  fade: "fade",
  dissolve: "dissolve",
  wipe_left: "wipeleft",
  wipe_right: "wiperight",
  wipe_up: "wipeup",
  wipe_down: "wipedown",
  slide_left: "slideleft",
  slide_right: "slideright",
  zoom_in: "squeezev",
  none: "fade",
};

/**
 * Compõe múltiplos vídeos em um único vídeo com transições.
 * Tudo roda no browser via FFmpeg.wasm.
 */
export async function composeVideos(
  files: File[],
  durations: number[],
  settings: CompositionSettings,
  onProgress?: (progress: number, message: string) => void
): Promise<Blob> {
  if (files.length === 0) throw new Error("Nenhum vídeo fornecido");
  if (files.length > 10) throw new Error("Máximo de 10 vídeos");

  onProgress?.(5, "Inicializando motor de vídeo...");
  const ffmpeg = await loadFFmpeg((msg) => onProgress?.(8, msg));

  const td = settings.transition_duration;
  const ffTransition = TRANSITION_MAP[settings.transition_type] || "fade";

  // Determinar resolução de saída - usar valores menores para performance no browser
  const outW = Math.min(settings.output_width, 720);
  const outH = Math.min(
    settings.output_height,
    Math.round((720 / settings.output_width) * settings.output_height)
  );

  // Calcular durações ideais
  const clipDurations = calculateClipDurations(durations, settings.target_duration, td);

  // 1. Carregar cada vídeo no sistema de arquivos virtual
  onProgress?.(10, "Carregando vídeos na memória...");
  for (let i = 0; i < files.length; i++) {
    const ext = files[i].name.split(".").pop() || "mp4";
    await ffmpeg.writeFile(`input_${i}.${ext}`, await fetchFile(files[i]));
    onProgress?.(
      10 + Math.round((i / files.length) * 15),
      `Carregando vídeo ${i + 1} de ${files.length}...`
    );
  }

  // 2. Normalizar cada clipe (resolução, duração, codec)
  onProgress?.(25, "Normalizando vídeos...");
  for (let i = 0; i < files.length; i++) {
    const ext = files[i].name.split(".").pop() || "mp4";
    const dur = clipDurations[i];

    await ffmpeg.exec([
      "-i", `input_${i}.${ext}`,
      "-t", dur.toFixed(2),
      "-vf", `scale=${outW}:${outH}:force_original_aspect_ratio=decrease,pad=${outW}:${outH}:(ow-iw)/2:(oh-ih)/2:black,setsar=1,fps=24`,
      "-c:v", "libx264",
      "-preset", "ultrafast",
      "-crf", "28",
      "-an",
      "-pix_fmt", "yuv420p",
      `-y`, `norm_${i}.mp4`,
    ]);

    onProgress?.(
      25 + Math.round((i / files.length) * 30),
      `Normalizando vídeo ${i + 1} de ${files.length}...`
    );
  }

  // 3. Compor com transições
  onProgress?.(60, "Aplicando transições...");

  if (files.length === 1) {
    // Apenas um vídeo - copiar diretamente
    await ffmpeg.exec([
      "-i", "norm_0.mp4",
      "-c", "copy",
      "-y", "output.mp4",
    ]);
  } else {
    // Construir filtro xfade para múltiplos vídeos
    const inputArgs: string[] = [];
    for (let i = 0; i < files.length; i++) {
      inputArgs.push("-i", `norm_${i}.mp4`);
    }

    // Obter durações reais dos arquivos normalizados
    // Para simplificar, usamos as durações calculadas
    const realDurations = clipDurations;

    let filterComplex = "";

    if (files.length === 2) {
      const offset = Math.max(0, realDurations[0] - td);
      filterComplex = `[0:v][1:v]xfade=transition=${ffTransition}:duration=${td}:offset=${offset.toFixed(2)}[outv]`;
    } else {
      // Primeiro par
      let offset = Math.max(0, realDurations[0] - td);
      filterComplex = `[0:v][1:v]xfade=transition=${ffTransition}:duration=${td}:offset=${offset.toFixed(2)}[v01]`;

      for (let i = 2; i < files.length; i++) {
        offset += realDurations[i - 1] - td;
        offset = Math.max(0, offset);
        const prevLabel = i === 2 ? "v01" : `v${i - 1}`;
        const outLabel = i === files.length - 1 ? "outv" : `v${i}`;
        filterComplex += `;[${prevLabel}][${i}:v]xfade=transition=${ffTransition}:duration=${td}:offset=${offset.toFixed(2)}[${outLabel}]`;
      }
    }

    onProgress?.(70, "Renderizando vídeo final...");

    await ffmpeg.exec([
      ...inputArgs,
      "-filter_complex", filterComplex,
      "-map", "[outv]",
      "-c:v", "libx264",
      "-preset", "ultrafast",
      "-crf", "26",
      "-pix_fmt", "yuv420p",
      "-movflags", "+faststart",
      "-y", "output.mp4",
    ]);
  }

  onProgress?.(90, "Finalizando...");

  // 4. Ler o arquivo de saída
  const data = await ffmpeg.readFile("output.mp4");

  // 5. Limpar arquivos temporários
  for (let i = 0; i < files.length; i++) {
    const ext = files[i].name.split(".").pop() || "mp4";
    try {
      await ffmpeg.deleteFile(`input_${i}.${ext}`);
      await ffmpeg.deleteFile(`norm_${i}.mp4`);
    } catch {
      // Ignorar erros de limpeza
    }
  }
  try {
    await ffmpeg.deleteFile("output.mp4");
  } catch {
    // Ignorar
  }

  onProgress?.(100, "Concluído!");

  // Converter Uint8Array para Blob
  const blob = new Blob([data], { type: "video/mp4" });
  return blob;
}

/**
 * Lista as transições disponíveis (local, sem API)
 */
export function getAvailableTransitions() {
  return [
    { id: "fade", name: "Fade", description: "Transição suave com fade in/out" },
    { id: "dissolve", name: "Dissolve", description: "Dissolução gradual entre clipes" },
    { id: "wipe_left", name: "Wipe Left", description: "Limpa da direita para a esquerda" },
    { id: "wipe_right", name: "Wipe Right", description: "Limpa da esquerda para a direita" },
    { id: "wipe_up", name: "Wipe Up", description: "Limpa de baixo para cima" },
    { id: "wipe_down", name: "Wipe Down", description: "Limpa de cima para baixo" },
    { id: "slide_left", name: "Slide Left", description: "Desliza para a esquerda" },
    { id: "slide_right", name: "Slide Right", description: "Desliza para a direita" },
    { id: "zoom_in", name: "Zoom Squeeze", description: "Efeito de zoom/squeeze vertical" },
    { id: "none", name: "Corte Seco", description: "Sem transição - corte direto" },
  ];
}

File: backend/video_processor.py
"""
Video Processor - Módulo de processamento de vídeos com FFmpeg.
Responsável por combinar múltiplos vídeos curtos com transições profissionais.
"""

import os
import subprocess
import json
import uuid
import shutil
from typing import List, Optional
from dataclasses import dataclass, field
from enum import Enum


class TransitionType(str, Enum):
    FADE = "fade"
    DISSOLVE = "dissolve"
    WIPE_LEFT = "wipe_left"
    WIPE_RIGHT = "wipe_right"
    WIPE_UP = "wipe_up"
    WIPE_DOWN = "wipe_down"
    SLIDE_LEFT = "slide_left"
    SLIDE_RIGHT = "slide_right"
    ZOOM_IN = "zoom_in"
    NONE = "none"


@dataclass
class VideoClip:
    path: str
    duration: float = 0.0
    width: int = 0
    height: int = 0
    fps: float = 30.0

    def __post_init__(self):
        if os.path.exists(self.path):
            self._probe()

    def _probe(self):
        """Obtém informações do vídeo usando ffprobe."""
        try:
            cmd = [
                "ffprobe", "-v", "quiet",
                "-print_format", "json",
                "-show_format", "-show_streams",
                self.path
            ]
            result = subprocess.run(cmd, capture_output=True, text=True, timeout=30)
            if result.returncode != 0:
                return

            info = json.loads(result.stdout)

            # Encontrar stream de vídeo
            for stream in info.get("streams", []):
                if stream.get("codec_type") == "video":
                    self.width = int(stream.get("width", 0))
                    self.height = int(stream.get("height", 0))
                    # Calcular FPS
                    r_frame_rate = stream.get("r_frame_rate", "30/1")
                    if "/" in r_frame_rate:
                        num, den = r_frame_rate.split("/")
                        self.fps = float(num) / float(den) if float(den) != 0 else 30.0
                    else:
                        self.fps = float(r_frame_rate)
                    break

            # Duração
            fmt = info.get("format", {})
            self.duration = float(fmt.get("duration", 0))
        except Exception as e:
            print(f"Erro ao analisar vídeo {self.path}: {e}")


@dataclass
class CompositionSettings:
    target_duration: float = 30.0  # Duração alvo em segundos
    transition_duration: float = 0.5  # Duração da transição em segundos
    transition_type: TransitionType = TransitionType.FADE
    output_width: int = 1080
    output_height: int = 1920  # Formato vertical (9:16)
    output_fps: int = 30
    output_format: str = "mp4"


class VideoComposer:
    """Classe principal para composição de vídeos."""

    def __init__(self, temp_dir: str = "temp", output_dir: str = "output"):
        self.temp_dir = temp_dir
        self.output_dir = output_dir
        os.makedirs(temp_dir, exist_ok=True)
        os.makedirs(output_dir, exist_ok=True)

    def _normalize_clip(self, clip: VideoClip, settings: CompositionSettings, index: int) -> str:
        """Normaliza um clipe: resolução, fps, codec."""
        output_path = os.path.join(self.temp_dir, f"normalized_{index}_{uuid.uuid4().hex[:8]}.mp4")

        # Filtro para escalar e adicionar padding para manter aspect ratio
        filter_complex = (
            f"scale={settings.output_width}:{settings.output_height}:"
            f"force_original_aspect_ratio=decrease,"
            f"pad={settings.output_width}:{settings.output_height}:(ow-iw)/2:(oh-ih)/2:black,"
            f"setsar=1,fps={settings.output_fps}"
        )

        cmd = [
            "ffmpeg", "-y", "-i", clip.path,
            "-vf", filter_complex,
            "-c:v", "libx264", "-preset", "fast",
            "-crf", "23",
            "-an",  # Remover áudio por simplicidade
            "-pix_fmt", "yuv420p",
            "-movflags", "+faststart",
            output_path
        ]

        result = subprocess.run(cmd, capture_output=True, text=True, timeout=120)
        if result.returncode != 0:
            raise RuntimeError(f"Erro ao normalizar vídeo {index}: {result.stderr}")

        return output_path

    def _trim_clip(self, clip_path: str, duration: float, index: int) -> str:
        """Corta um clipe para a duração especificada."""
        output_path = os.path.join(self.temp_dir, f"trimmed_{index}_{uuid.uuid4().hex[:8]}.mp4")

        cmd = [
            "ffmpeg", "-y", "-i", clip_path,
            "-t", str(duration),
            "-c:v", "libx264", "-preset", "fast",
            "-crf", "23",
            "-pix_fmt", "yuv420p",
            output_path
        ]

        result = subprocess.run(cmd, capture_output=True, text=True, timeout=120)
        if result.returncode != 0:
            raise RuntimeError(f"Erro ao cortar vídeo {index}: {result.stderr}")

        return output_path

    def _calculate_clip_durations(
        self, clips: List[VideoClip], settings: CompositionSettings
    ) -> List[float]:
        """Calcula a duração de cada clipe para atingir a duração alvo."""
        n = len(clips)
        if n == 0:
            return []

        # Tempo total das transições
        total_transition_time = (n - 1) * settings.transition_duration
        # Tempo disponível para conteúdo
        available_time = settings.target_duration + total_transition_time

        # Distribuir proporcionalmente baseado nas durações originais
        total_original = sum(c.duration for c in clips)

        if total_original <= 0:
            # Distribuir igualmente
            per_clip = available_time / n
            return [per_clip] * n

        durations = []
        for clip in clips:
            ratio = clip.duration / total_original
            clip_dur = ratio * available_time
            # Limitar entre 2s e duração original
            clip_dur = min(clip_dur, clip.duration)
            clip_dur = max(clip_dur, 2.0)
            durations.append(clip_dur)

        return durations

    def _build_xfade_filter(
        self,
        num_clips: int,
        durations: List[float],
        settings: CompositionSettings,
    ) -> str:
        """Constrói o filtro xfade para transições entre clipes."""

        transition_map = {
            TransitionType.FADE: "fade",
            TransitionType.DISSOLVE: "dissolve",
            TransitionType.WIPE_LEFT: "wipeleft",
            TransitionType.WIPE_RIGHT: "wiperight",
            TransitionType.WIPE_UP: "wipeup",
            TransitionType.WIPE_DOWN: "wipedown",
            TransitionType.SLIDE_LEFT: "slideleft",
            TransitionType.SLIDE_RIGHT: "slideright",
            TransitionType.ZOOM_IN: "squeezev",
            TransitionType.NONE: "fade",
        }

        ffmpeg_transition = transition_map.get(settings.transition_type, "fade")
        td = settings.transition_duration

        if num_clips == 1:
            return "[0:v]copy[outv]"

        filters = []
        offset = durations[0] - td

        if num_clips == 2:
            filters.append(
                f"[0:v][1:v]xfade=transition={ffmpeg_transition}:"
                f"duration={td}:offset={max(0, offset)}[outv]"
            )
        else:
            # Primeiro par
            filters.append(
                f"[0:v][1:v]xfade=transition={ffmpeg_transition}:"
                f"duration={td}:offset={max(0, offset)}[v01]"
            )

            for i in range(2, num_clips):
                offset += durations[i - 1] - td
                prev_label = f"v{str(i - 2).zfill(2)}{str(i - 1).zfill(2)}" if i == 2 else f"v{i - 1}"
                if i == 2:
                    prev_label = "v01"

                if i == num_clips - 1:
                    out_label = "outv"
                else:
                    out_label = f"v{i}"

                filters.append(
                    f"[{prev_label}][{i}:v]xfade=transition={ffmpeg_transition}:"
                    f"duration={td}:offset={max(0, offset)}[{out_label}]"
                )

        return ";".join(filters)

    def compose(
        self,
        video_paths: List[str],
        settings: Optional[CompositionSettings] = None,
        job_id: Optional[str] = None,
    ) -> str:
        """
        Compõe múltiplos vídeos em um único vídeo final com transições.

        Args:
            video_paths: Lista de caminhos dos vídeos de entrada
            settings: Configurações da composição
            job_id: ID único do job

        Returns:
            Caminho do vídeo final
        """
        if settings is None:
            settings = CompositionSettings()

        if not job_id:
            job_id = uuid.uuid4().hex[:12]

        if len(video_paths) == 0:
            raise ValueError("Nenhum vídeo fornecido")

        if len(video_paths) > 10:
            raise ValueError("Máximo de 10 vídeos permitidos")

        # 1. Carregar e analisar clipes
        clips = [VideoClip(path=p) for p in video_paths]
        for i, clip in enumerate(clips):
            if clip.duration <= 0:
                raise ValueError(f"Vídeo {i + 1} não pôde ser analisado ou está vazio")

        # 2. Calcular durações
        durations = self._calculate_clip_durations(clips, settings)

        # 3. Normalizar clipes
        normalized_paths = []
        for i, clip in enumerate(clips):
            norm_path = self._normalize_clip(clip, settings, i)
            # Cortar se necessário
            norm_clip = VideoClip(path=norm_path)
            if norm_clip.duration > durations[i]:
                trimmed = self._trim_clip(norm_path, durations[i], i)
                normalized_paths.append(trimmed)
            else:
                normalized_paths.append(norm_path)

        # Recalcular durações reais após normalização
        actual_durations = []
        for p in normalized_paths:
            c = VideoClip(path=p)
            actual_durations.append(c.duration)

        # 4. Construir filtro de transição
        filter_complex = self._build_xfade_filter(
            len(normalized_paths), actual_durations, settings
        )

        # 5. Gerar vídeo final
        output_path = os.path.join(self.output_dir, f"composed_{job_id}.mp4")

        cmd = ["ffmpeg", "-y"]
        for p in normalized_paths:
            cmd.extend(["-i", p])

        cmd.extend([
            "-filter_complex", filter_complex,
            "-map", "[outv]",
            "-c:v", "libx264", "-preset", "fast",
            "-crf", "20",
            "-pix_fmt", "yuv420p",
            "-movflags", "+faststart",
            output_path
        ])

        result = subprocess.run(cmd, capture_output=True, text=True, timeout=300)
        if result.returncode != 0:
            raise RuntimeError(f"Erro na composição final: {result.stderr}")

        # 6. Limpar arquivos temporários
        for p in normalized_paths:
            try:
                os.remove(p)
            except OSError:
                pass

        return output_path

    def get_video_info(self, video_path: str) -> dict:
        """Retorna informações sobre um vídeo."""
        clip = VideoClip(path=video_path)
        return {
            "duration": round(clip.duration, 2),
            "width": clip.width,
            "height": clip.height,
            "fps": round(clip.fps, 2),
        }

    def generate_thumbnail(self, video_path: str, output_path: str, time: float = 0.5) -> str:
        """Gera uma thumbnail de um vídeo."""
        cmd = [
            "ffmpeg", "-y", "-i", video_path,
            "-ss", str(time),
            "-vframes", "1",
            "-vf", "scale=320:-1",
            output_path
        ]
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=30)
        if result.returncode != 0:
            raise RuntimeError(f"Erro ao gerar thumbnail: {result.stderr}")
        return output_path

    def cleanup_job(self, job_id: str):
        """Limpa arquivos temporários de um job."""
        for f in os.listdir(self.temp_dir):
            if job_id in f:
                try:
                    os.remove(os.path.join(self.temp_dir, f))
                except OSError:
                    pass
